### ISSUE TITLE
debugger command "|x" swallows all output

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#6582]
Migrated from [rt.perl.org#22758](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=22758) (status was 'resolved')

Searchable as RT22758$

**Comment by p5pRT** [comment]
### From Ulrich.Windl@rz.uni-regensburg.de

#### Created by windl@elf

This is a bug report for perl from windl@&#8203;elf\,
generated with the help of perlbug 1\.34 running under perl v5\.8\.0\.


\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
The debugger command "x" works as expected\, but "|x" shows no output at all&#8203;:
&nbsp;  DB\<4> x $tagfilepos
0  HASH\(0x8529ef8\)
&nbsp;   258 => 9084218
&nbsp;   269 => 9084224
&nbsp;   270 => 9083996
&nbsp;   271 => 9084002
&nbsp;   272 => 9084014
&nbsp;   273 => 9084332
&nbsp;   279 => 9084468
&nbsp;   282 => 9084604
&nbsp;   283 => 9084612
&nbsp;   305 => 9084020
&nbsp;   306 => 9084036
&nbsp;   315 => 9084056
&nbsp;   33432 => 9087764
&nbsp;   33434 => 9087922
&nbsp;   33437 => 9087930
&nbsp;   34675 => 9084620
&nbsp;   36867 => 9087938
&nbsp;   36868 => 9087958
&nbsp;   37377 => 9087978
&nbsp;   37378 => 9087986
&nbsp;   37381 => 9087994
&nbsp;   37386 => 9088002
&nbsp;   37500 => 9088010
&nbsp;   37510 => 9088028
&nbsp;   41493 => 9088048
&nbsp;   41989 => 9088056
&nbsp;  DB\<5> |x $tagfilepos
&nbsp;  DB\<6> 

\(interestingly\, this doesn't happen for trivial cases\)&#8203;:
windl@&#8203;elf&#8203;:\~/Perl> perl \-d \-e 1

Loading DB routines from perl5db\.pl version 1\.19
Editor support available\.

Enter h or \`h h' for help\, or \`man perldebug' for more help\.

main&#8203;::\(\-e&#8203;:1\)&#8203;:   1
&nbsp;  DB\<1> @&#8203;a = \(0\.\.9\)

&nbsp;  DB\<2> x @&#8203;a
0  0
1  1
2  2
3  3
4  4
5  5
6  6
7  7
8  8
9  9
&nbsp;  DB\<3> |x @&#8203;a
\(works\)

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=medium

This perlbug was built using Perl v5.8.0 - Thu Mar 13 22:21:38 UTC 2003
It is being executed now by  Perl v5.8.0 - Thu Mar 13 22:14:08 UTC 2003.

Site configuration information for perl v5.8.0:

Configured by root at Thu Mar 13 22:14:08 UTC 2003.

Summary of my perl5 (revision 5.0 version 8 subversion 0) configuration:
  Platform:
    osname=linux, osvers=2.4.20, archname=i586-linux-thread-multi
    uname='linux d20 2.4.20 #1 smp thu oct 10 18:10:26 utc 2002 i686 unknown unknown 
gnulinux '
    config_args='-ds -e -Dprefix=/usr -Dusethreads -Di_db -Di_dbm -Di_ndbm -Di_gdbm -
Duseshrplib=true'
    hint=recommended, useposix=true, d_sigaction=define
    usethreads=define use5005threads=undef useithreads=define usemultiplicity=define
    useperlio=define d_sfio=undef uselargefiles=define usesocks=undef
    use64bitint=undef use64bitall=undef uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='cc', ccflags ='-D_REENTRANT -D_GNU_SOURCE -fno-strict-aliasing -
D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64',
    optimize='-O3 --pipe',
    cppflags='-D_REENTRANT -D_GNU_SOURCE -fno-strict-aliasing'
    ccversion='', gccversion='3.3 20030226 (prerelease) (SuSE Linux)', gccosandvers=''
    intsize=4, longsize=4, ptrsize=4, doublesize=8, byteorder=1234
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=12
    ivtype='long', ivsize=4, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=8
    alignbytes=4, prototype=define
  Linker and Libraries:
    ld='cc', ldflags =''
    libpth=/lib /usr/lib /usr/local/lib
    libs=-lnsl -ldl -lm -lpthread -lc -lcrypt -lutil
    perllibs=-lnsl -ldl -lm -lpthread -lc -lcrypt -lutil
    libc=, so=so, useshrplib=true, libperl=libperl.so
    gnulibc_version='2.3.2'
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-rdynamic -Wl,-
rpath,/usr/lib/perl5/5.8.0/i586-linux-thread-multi/CORE'
    cccdlflags='-fPIC', lddlflags='-shared'

Locally applied patches:
    


@INC for perl v5.8.0:
    /usr/lib/perl5/5.8.0/i586-linux-thread-multi
    /usr/lib/perl5/5.8.0
    /usr/lib/perl5/site_perl/5.8.0/i586-linux-thread-multi
    /usr/lib/perl5/site_perl/5.8.0
    /usr/lib/perl5/site_perl
    .


Environment for perl v5.8.0:
    HOME=/home/windl
    LANG=de_DE@euro
    LANGUAGE (unset)
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    
PATH=/usr/local/bin:/usr/bin:/usr/X11R6/bin:/bin:/var/lib/dosemu:/usr/games:/opt/gnome2
/bin:/opt/gnome/bin:/opt/kde3/bin:/usr/openwin/bin:/usr/lib/SmallEiffel/bin:/usr/lib/ja
va/bin:/opt/gnome/bin
    PERL_BADLANG (unset)
    SHELL=/bin/bash


```
</details>

**Comment by p5pRT** [comment]
### From @schwern

> \[Ulrich\.Windl@&#8203;rz\.uni\-regensburg\.de \- Sun Jun 22 23&#8203;:53&#8203;:28 2003\]&#8203;:
> 
> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> The debugger command "x" works as expected\, but "|x" shows no output
> at all&#8203;:

I am unable to reproduce this problem\.  What pager does the debugger
think you're using?  You can find out with this&#8203;:

$ perl \-de 1

Loading DB routines from perl5db\.pl version 1\.28
Editor support available\.

Enter h or \`h h' for help\, or \`man perldebug' for more help\.

main&#8203;::\(\-e&#8203;:1\)&#8203;:   1
&nbsp;  DB\<1> print DB&#8203;::pager
|/usr/bin/less

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From Ulrich.Windl@rz.uni-regensburg.de

On 12 Jul 2005 at 16&#8203;:19\, Michael G Schwern via RT wrote&#8203;:

> > \[Ulrich\.Windl@&#8203;rz\.uni\-regensburg\.de \- Sun Jun 22 23&#8203;:53&#8203;:28 2003\]&#8203;:
> > 
> > \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> > The debugger command "x" works as expected\, but "|x" shows no output
> > at all&#8203;:
> 
> I am unable to reproduce this problem\.  What pager does the debugger
> think you're using?  You can find out with this&#8203;:
> 
> $ perl \-de 1
> 
> Loading DB routines from perl5db\.pl version 1\.28
> Editor support available\.
> 
> Enter h or \`h h' for help\, or \`man perldebug' for more help\.
> 
> main&#8203;::\(\-e&#8203;:1\)&#8203;:   1
>   DB\<1> print DB&#8203;::pager
> |/usr/bin/less
> 

main&#8203;::\(\-e&#8203;:1\)&#8203;:   1
&nbsp;  DB\<1> print DB&#8203;::pager
|less

So maybe it's the missting fully qualified name\. Shouldn't perl print some 
diagnostics if the pager cannot be found?

BTW&#8203;: I tried \(not sure which configuration I had reported the problem\)&#8203;:

mailgate&#8203;:/var/tmp \# PATH= /usr/bin/perl \-de 1

Loading DB routines from perl5db\.pl version 1\.27
Editor support available\.

Enter h or \`h h' for help\, or \`man perldebug' for more help\.

main&#8203;::\(\-e&#8203;:1\)&#8203;:   1
&nbsp;  DB\<1> @&#8203;a=\(1 \.\. 17\)

&nbsp;  DB\<2> x @&#8203;a
0  1
1  2
2  3
3  4
4  5
5  6
6  7
7  8
8  9
9  10
10  11
11  12
12  13
13  14
14  15
15  16
16  17
&nbsp;  DB\<3> |x @&#8203;a

\[1\]\+  Stopped                 PATH= /usr/bin/perl \-de 1
\(Perl seems to hang here\)

Regards\,
Ulrich

**Comment by p5pRT** [comment]
### From @richardfoley

This is a known bug which seems to hit the debugger periodically\, and I
think is related to the particular combination of modules installed on a
given system\, in particular Term&#8203;::ReadLine&#8203;::Gnu\.  There is a simple
workaround which involves setting\, \(if I remember correctly\)\,
PERLDB\_RL=1\.    You might also try installing Term&#8203;::ReadLine&#8203;::Perl\, I
think that solves it too\.  Basically\, I think PERL\_RL just has to be
non\-zero\.  Something like this&#8203;:

&nbsp;   $> PERL\_RL=1 perl \-d \.\.\.

I realize the answer is a bit vague\, but my reference is elsewhere\,
among other things ;\-\)

\-\- 
Ciao
Richard Foley
Ciao \- shorter than aufwiedersehen

http&#8203;://www\.oreilly\.com/catalog/perldebugpr/ 

\-\-\-\-\-Original Message\-\-\-\-\-
> Date&#8203;: Mon\,  1 Aug 2005 11&#8203;:32&#8203;:41 \+0200
> Subject&#8203;: Re&#8203;: \[perl \#22758\] debugger command "|x" swallows all output 
> From&#8203;: "Ulrich Windl" \<ulrich\.windl@&#8203;rz\.uni\-regensburg\.de>
> To&#8203;: "Michael G Schwern via RT" \<perlbug\-followup@&#8203;perl\.org>

> On 12 Jul 2005 at 16&#8203;:19\, Michael G Schwern via RT wrote&#8203;:
> 
> > > \[Ulrich\.Windl@&#8203;rz\.uni\-regensburg\.de \- Sun Jun 22 23&#8203;:53&#8203;:28 2003\]&#8203;: 
> > > \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> > > The debugger command "x" works as expected\, but "|x" shows no
> > > output at all&#8203;:
> > > 
> > 
> > I am unable to reproduce this problem\.  What pager does the debugger
> > think you're using?  You can find out with this&#8203;:
> > 
> > $ perl \-de 1
> > 
> > Loading DB routines from perl5db\.pl version 1\.28
> > Editor support available\.
> > 
> > Enter h or \`h h' for help\, or \`man perldebug' for more help\.
> > 
> > main&#8203;::\(\-e&#8203;:1\)&#8203;:   1
> > DB\<1> print DB&#8203;::pager
> > |/usr/bin/less
> > 
> > 
> 
> main&#8203;::\(\-e&#8203;:1\)&#8203;:   1
> DB\<1> print DB&#8203;::pager
> |less
> 
> So maybe it's the missting fully qualified name\. Shouldn't perl print
> some diagnostics if the pager cannot be found?
> 
> BTW&#8203;: I tried \(not sure which configuration I had reported the
> problem\)&#8203;: 
> mailgate&#8203;:/var/tmp \# PATH= /usr/bin/perl \-de 1
> 
> Loading DB routines from perl5db\.pl version 1\.27
> Editor support available\.
> 
> Enter h or \`h h' for help\, or \`man perldebug' for more help\.
> 
> main&#8203;::\(\-e&#8203;:1\)&#8203;:   1
> DB\<1> @&#8203;a=\(1 \.\. 17\)
> 
> DB\<2> x @&#8203;a
> 0  1
> 1  2
> 2  3
> 3  4
> 4  5
> 5  6
> 6  7
> 7  8
> 8  9
> 9  10
> 10  11
> 11  12
> 12  13
> 13  14
> 14  15
> 15  16
> 16  17
> DB\<3> |x @&#8203;a
> 
> \[1\]\+  Stopped                 PATH= /usr/bin/perl \-de 1
> \(Perl seems to hang here\)
> 
> Regards\,
> Ulrich
> 
> 
>

**Comment by p5pRT** [comment]
### From @richardfoley

That should be PERL\_RL=1

Please ignore the perldb\_rl variable erroneously mentioned in the above
post\.

\-\- 
Ciao
Richard Foley
Ciao \- shorter than aufwiedersehen

http&#8203;://www\.oreilly\.com/catalog/perldebugpr/ 

\-\-\-\-\-Original Message\-\-\-\-\-
> Date&#8203;: Tue\,  2 Aug 2005 08&#8203;:05&#8203;:57 \+0200
> Subject&#8203;: Re&#8203;: Re&#8203;: \[perl \#22758\] debugger command "|x" swallows all
> output 
> From&#8203;: "Richard\.Foley@&#8203;t\-online\.de" \<Richard\.Foley@&#8203;t\-online\.de>
> To&#8203;: "Ulrich Windl" \<ulrich\.windl@&#8203;rz\.uni\-regensburg\.de>\, "Michael G
> Schwern via RT" \<perlbug\-followup@&#8203;perl\.org>

> This is a known bug which seems to hit the debugger periodically\, and
> I think is related to the particular combination of modules installed
> on a given system\, in particular Term&#8203;::ReadLine&#8203;::Gnu\.  There is a
> simple workaround which involves setting\, \(if I remember correctly\)\,
> PERLDB\_RL=1\.    You might also try installing Term&#8203;::ReadLine&#8203;::Perl\, I
> think that solves it too\.  Basically\, I think PERL\_RL just has to be
> non\-zero\.  Something like this&#8203;:
> 
> $> PERL\_RL=1 perl \-d \.\.\.
> 
> I realize the answer is a bit vague\, but my reference is elsewhere\,
> among other things ;\-\)
> 
> \-\-
> Ciao
> Richard Foley
> Ciao \- shorter than aufwiedersehen
> 
> http&#8203;://www\.oreilly\.com/catalog/perldebugpr/
> 
> \-\-\-\-\-Original Message\-\-\-\-\-
> 
> > Date&#8203;: Mon\,  1 Aug 2005 11&#8203;:32&#8203;:41 \+0200
> > Subject&#8203;: Re&#8203;: \[perl \#22758\] debugger command "|x" swallows all output
> > From&#8203;: "Ulrich Windl" \<ulrich\.windl@&#8203;rz\.uni\-regensburg\.de>
> > To&#8203;: "Michael G Schwern via RT" \<perlbug\-followup@&#8203;perl\.org>
> > 
> 
> > On 12 Jul 2005 at 16&#8203;:19\, Michael G Schwern via RT wrote&#8203;:
> > 
> > > > \[Ulrich\.Windl@&#8203;rz\.uni\-regensburg\.de \- Sun Jun 22 23&#8203;:53&#8203;:28 2003\]&#8203;:
> > > >


---
> > > > The debugger command "x" works as expected\, but "|x" shows no
> > > > output at all&#8203;:
> > > > 
> > > > 
> > > 
> > > I am unable to reproduce this problem\.  What pager does the
> > > debugger think you're using?  You can find out with this&#8203;:
> > > 
> > > $ perl \-de 1
> > > 
> > > Loading DB routines from perl5db\.pl version 1\.28
> > > Editor support available\.
> > > 
> > > Enter h or \`h h' for help\, or \`man perldebug' for more help\.
> > > 
> > > main&#8203;::\(\-e&#8203;:1\)&#8203;:   1
> > > DB\<1> print DB&#8203;::pager
> > > |/usr/bin/less
> > > 
> > > 
> > > 
> > 
> > main&#8203;::\(\-e&#8203;:1\)&#8203;:   1
> > DB\<1> print DB&#8203;::pager
> > |less
> > 
> > So maybe it's the missting fully qualified name\. Shouldn't perl
> > print some diagnostics if the pager cannot be found?
> > 
> > BTW&#8203;: I tried \(not sure which configuration I had reported the
> > problem\)&#8203;:
> > mailgate&#8203;:/var/tmp \# PATH= /usr/bin/perl \-de 1
> > 
> > Loading DB routines from perl5db\.pl version 1\.27
> > Editor support available\.
> > 
> > Enter h or \`h h' for help\, or \`man perldebug' for more help\.
> > 
> > main&#8203;::\(\-e&#8203;:1\)&#8203;:   1
> > DB\<1> @&#8203;a=\(1 \.\. 17\)
> > 
> > DB\<2> x @&#8203;a
> > 0  1
> > 1  2
> > 2  3
> > 3  4
> > 4  5
> > 5  6
> > 6  7
> > 7  8
> > 8  9
> > 9  10
> > 10  11
> > 11  12
> > 12  13
> > 13  14
> > 14  15
> > 15  16
> > 16  17
> > DB\<3> |x @&#8203;a
> > 
> > \[1\]\+  Stopped                 PATH= /usr/bin/perl \-de 1
> > \(Perl seems to hang here\)
> > 
> > Regards\,
> > Ulrich
> > 
> > 
> > 
> > 
> 
> 
> 
>

**Comment by p5pRT** [comment]
### From @kevindawson

I think this can be closed\, also \#37781


$ perlbrew use perl\-5\.10\.1
$ perl \-d \-e 1

Loading DB routines from perl5db\.pl version 1\.32
Editor support available\.

Enter h or \`h h' for help\, or \`man perldebug' for more help\.

main&#8203;::\(\-e&#8203;:1\)&#8203;:    1
&nbsp;  DB\<1> @&#8203;a=\(1\.\.9\)

&nbsp;  DB\<2> x @&#8203;a
0  1
1  2
2  3
3  4
4  5
5  6
6  7
7  8
8  9
&nbsp;  DB\<3> |x @&#8203;a
&nbsp;  DB\<4> 

It works has just not been tested recently

**Comment by p5pRT** [comment]
### From [Unknown Contact. See original ticket]

I think this can be closed\, also \#37781


$ perlbrew use perl\-5\.10\.1
$ perl \-d \-e 1

Loading DB routines from perl5db\.pl version 1\.32
Editor support available\.

Enter h or \`h h' for help\, or \`man perldebug' for more help\.

main&#8203;::\(\-e&#8203;:1\)&#8203;:    1
&nbsp;  DB\<1> @&#8203;a=\(1\.\.9\)

&nbsp;  DB\<2> x @&#8203;a
0  1
1  2
2  3
3  4
4  5
5  6
6  7
7  8
8  9
&nbsp;  DB\<3> |x @&#8203;a
&nbsp;  DB\<4> 

It works has just not been tested recently

**Comment by p5pRT** [comment]
### From @jkeenan

On Fri Aug 23 09&#8203;:20&#8203;:55 2013\, bowtie wrote&#8203;:
> I think this can be closed\, also \#37781
> 
> 
> $ perlbrew use perl\-5\.10\.1
> $ perl \-d \-e 1
> 
> Loading DB routines from perl5db\.pl version 1\.32
> Editor support available\.
> 
> Enter h or \`h h' for help\, or \`man perldebug' for more help\.
> 
> main&#8203;::\(\-e&#8203;:1\)&#8203;:    1
>   DB\<1> @&#8203;a=\(1\.\.9\)
> 
>   DB\<2> x @&#8203;a
> 0  1
> 1  2
> 2  3
> 3  4
> 4  5
> 5  6
> 6  7
> 7  8
> 8  9
>   DB\<3> |x @&#8203;a
>   DB\<4> 
> 
> It works has just not been tested recently

Agreed\.  According to pod/perldebug\.pod\, "|dbcmd" means&#8203;: Run 
the debugger command\, piping DB&#8203;::OUT into your current pager\.

I tried both data sets provided in the OP and confirmed that "|dbcmd"
activated my pager \(/usr/bin/less \-R\)\.

Thanks for the update\.  Resolving ticket\.

**Comment by p5pRT** [comment]
@jkeenan - Status changed from 'open' to 'resolved'
